feat: job feeder and monitor

ceholden opened the pull request:
## What I am changing

This PR adds,

* A "Granule Tracker Service" to keep track of our progress in submitting granules from our inventory files to our AWS Batch cluster for processing.
* A "Granule Logger Service" to handle logging successes and failures to S3 for persistent state tracking of our job outcomes.
* An "AWS Batch client" to help take care of `ListJob` and `SubmitJob` actions that will be used by the "queue feeder" and "requeuer" (and to facilitate unit testing)
* Other common domain concepts (our "granule ID", processing outcomes, final outcomes, and our processing "event")
* Lambda function to convert the inventories into Parquet for easy iteration
* Queue feeder that reads our inventory Parquet files and submits work
* Queue monitor that logs outcomes and retries failures

## How I did it

* Add common components (mostly in `common/models.py`)
* Add granule tracker
* Realize reading from the raw inventory files would be terrible, so wrote the converter to Parquet
* Added granule logger
* Wired up queue monitor
* Wired up queue feeder
* Lots of unit testing

## How you can test it

```
scripts/test
```

chuckwondo reviewed (commented):
Great infrastructure work! Thanks for pairing up with me to walk me through everything.

I don't see any issues with the infrastructure itself. My comments all relate to sorting out Python packaging issues and making the static typing more robust.

If you choose to apply my recommendations, you will also need to add the following files (both empty):

- `lambdas/py.typed`
- `lambdas/__init__.py`

Please also add the following to the "Deployment" section of the README.md file:

```
This project uses the AWS CDK.  See
[Getting Started with the AWS CDK](https://docs.aws.amazon.com/cdk/v2/guide/getting-started.html)
for installation instructions.
```

CAUTION: Although I made sure all of my suggestions work with pytest, mypy, and ruff, I did not yet manage to get the cdk build to complete (`uv run cdk synth` has been running for quite a while since I created/populated a `.env` file).

IMPORTANT: I see that some of the GitHub env vars need to be renamed to match recent name changes. I also see that the env var for the "protected" bucket has "PRIVATE" in the name, rather than "PROTECTED," which is perhaps slightly confusing.

chuckwondo commented on cdk/stack.py:
```diff
@@ -13,16 +16,72 @@
 from settings import StackSettings
 
 LAMBDA_EXCLUDE = [
+    ".git",
+    ".github",
```
```suggestion
    ".coverage",
    ".git*",
    "**/__pycache__",
```

chuckwondo reviewed (commented):
I finally managed to get `uv run cdk synth` to work (it was hanging because Docker was hanging, so I had to completely shut down Docker and restart it).

Anywho, I managed to uncover a few issues that my suggestions from my previous set of comment introduced. The suggestions here rectify the issues. Taken together, I believe all of the suggestions allow for improved dev experience in conjunction with the desired deployment structure.

ceholden commented on pyproject.toml:
```diff
@@ -5,35 +5,68 @@ description = "Orchestration of the historical processing for the HLS-VI product
+]
+
+[tool.uv]
+default-groups = ["dev"]
```
I don't think we want to use setuptools (prefer hatch w/ uv) but it does seem like the contents of `lambdas/*` is missing from the package list

ceholden replied:
I think I addressed this via,

```toml
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[tool.hatch.build.targets.wheel]
packages = ["src/common", "src/job_feeder", "src/job_monitor", "src/queue_feeder"]

[tool.uv]
default-groups = ["dev"]
package = true
```

plus `git mv lambdas src`. This tells `uv` to consider this a package and to identify the directories within `src/` as packages.

I checked the bdist and it looked good,
```
➜ uv build --wheel
Building wheel...
Successfully built dist/hls_vi_historical_orchestration-0.1.0-py3-none-any.whl

➜ unzip -l dist/hls_vi_historical_orchestration-0.1.0-py3-none-any.whl
Archive:  dist/hls_vi_historical_orchestration-0.1.0-py3-none-any.whl
  Length      Date    Time    Name
---------  ---------- -----   ----
      738  02-02-2020 00:00   common/__init__.py
     4014  02-02-2020 00:00   common/aws_batch.py
     7664  02-02-2020 00:00   common/granule_logger.py
     8053  02-02-2020 00:00   common/granule_tracker.py
     3692  02-02-2020 00:00   common/models.py
        0  02-02-2020 00:00   job_monitor/__init__.py
     2477  02-02-2020 00:00   job_monitor/handler.py
        0  02-02-2020 00:00   queue_feeder/__init__.py
     1893  02-02-2020 00:00   queue_feeder/handler.py
      976  02-02-2020 00:00   hls_vi_historical_orchestration-0.1.0.dist-info/METADATA
       87  02-02-2020 00:00   hls_vi_historical_orchestration-0.1.0.dist-info/WHEEL
     1061  02-02-2020 00:00   hls_vi_historical_orchestration-0.1.0.dist-info/licenses/LICENSE
     1098  02-02-2020 00:00   hls_vi_historical_orchestration-0.1.0.dist-info/RECORD
---------                     -------
    31753                     13 files
```

ceholden commented on lambdas/common/aws_batch.py:
```diff
@@ -0,0 +1,123 @@
+
+import boto3
+
+from common.models import GranuleProcessingEvent, JobOutcome
```
I wasn't intending for `lambdas` to be a Python package, sort of like you wouldn't do that with the "src" folder (`from src.common import thing`)

I'm guessing this suggestion relates to your editor tooling (autocomplete/etc)? As I said in the last PR, I think of each subfolder here as a separate package in a workspace so I'm hoping to avoid `from lambdas.<submodule> import <thing>`

chuckwondo replied:
Ok, I see now.

ceholden commented on tests/common/test_granule_logger.py:
```diff
@@ -0,0 +1,112 @@
+"""Tests for `common.granule_logger`"""
+
+from typing import cast
```
Awesome! Thanks for the suggestions to fix this

ceholden commented:
>IMPORTANT: I see that some of the GitHub env vars need to be renamed to match recent name changes. I also see that the env var for the "protected" bucket has "PRIVATE" in the name, rather than "PROTECTED," which is perhaps slightly confusing.

Fixed this! Thanks for finding this, the two words had collapsed into the same letters in my mind

ceholden commented on lambdas/inventory_converter/handler.py:
```diff
@@ -0,0 +1,183 @@
+import pyarrow.parquet as pq
+
+logger = logging.getLogger(__name__)
+if logger.hasHandlers():
+    logger.setLevel(logging.INFO)
+else:
+    logging.basicConfig(level=logging.INFO)
```
The Lambda function should already have a handler setup but I wanted to configure one so I could see output for local/dev testing

chuckwondo replied:
I figured as much. Thanks for clarifying.

ceholden commented on pyproject.toml:
```diff
@@ -5,35 +5,68 @@ description = "Orchestration of the historical processing for the HLS-VI product
 strict = true
 show_error_code_links = true
 show_error_context = true
+no_namespace_packages = true
```
I need a linter for unrecognized settings 😭 thanks for catching this! I've removed it

ceholden commented on cdk/stack.py:
```diff
@@ -13,16 +16,72 @@
+            "pip install uv",
+            "export UV_CACHE_DIR=/tmp",
+            f"cd {UV_ASSET_REQUIREMENTS}",
+            f"uv export {groups_arg} --frozen --no-dev --no-default-groups --no-editable -o {input_dir}/requirements.txt",
```
Thanks! This was helpful after moving to have `uv` actually manage this as a project with packages

ceholden commented on lambdas/common/models.py:
```diff
@@ -1,38 +1,133 @@
+            self.SUCCESS: ProcessingOutcome.SUCCESS,
+            self.FAILURE_RETRYABLE: ProcessingOutcome.FAILURE,
+            self.FAILURE_NONRETRYABLE: ProcessingOutcome.FAILURE,
+        }[self]  # type: ignore[index]
```
Nice idea with the `auto(), ProcessingOutcome.[value]`!

I am leaning towards defining the `processing_outcome` as a property to try to avoid anything related to `__new__` 🙈,

e.g.,
```python
@unique
class JobOutcome(Enum):
    SUCCESS = auto(), ProcessingOutcome.SUCCESS
    FAILURE_RETRYABLE = auto(), ProcessingOutcome.FAILURE
    FAILURE_NONRETRYABLE = auto(), ProcessingOutcome.FAILURE

    @property
    def processing_outcome(self) -> ProcessingOutcome:
        return self.value[1]
```

chuckwondo replied:
If that does the trick, that's certainly simpler.

Perhaps even simpler would be to drop the `@unique` constraint (and thus also the need for `auto`), unless there's something requiring uniqueness:

```python
class JobOutcome(Enum):
    SUCCESS = ProcessingOutcome.SUCCESS
    FAILURE_RETRYABLE = ProcessingOutcome.FAILURE
    FAILURE_NONRETRYABLE = ProcessingOutcome.FAILURE

    @property
    def processing_outcome(self) -> ProcessingOutcome:
        return self.value
```

I tried this earlier in my experimentations while reviewing this PR, but tests were failing. However, I discovered that the tests were failing equality checks for JobOutcome instances. Even though they values being compared looked identical, they were not comparing as equal, so I was utterly puzzled.

I later realized that this was due to the whole packaging structure, and that somehow the "expected" outcome instance created in the test was likely an instance of JobOutcome loaded via a different module/namespace than the "actual" outcome instanced produced by the code under test, so even though they looked identical, they were technically not because the classes were not the same due to namespace whackiness.

However, once I sorted out the packaging (the comments you won't incorporate, since you will take a slightly different approach) I didn't go back to see if the simplification above would finally work.

I'd be curious to know if it would work with your packaging adjustments in place, or is there something requiring uniqueness?

ceholden replied:
I was surprised by this because I thought enums were unique by the name, but it's actually the value. If the values are the same, the enum will only have 1 of those members,

```python
In [1]: from enum import Enum

In [2]: from common.models import ProcessingOutcome

In [3]: class JobOutcome(Enum):
   ...:     SUCCESS = ProcessingOutcome.SUCCESS
   ...:     FAILURE_RETRYABLE = ProcessingOutcome.FAILURE
   ...:     FAILURE_NONRETRYABLE = ProcessingOutcome.FAILURE
   ...:

In [4]: list(JobOutcome)
Out[4]:
[<JobOutcome.SUCCESS: <ProcessingOutcome.SUCCESS: 1>>,
 <JobOutcome.FAILURE_RETRYABLE: <ProcessingOutcome.FAILURE: 2>>]
```

The `@unique` is actually checking that there is only 1 enum name pointing to the same value to help make sure we don't shadow one of the entries with another

chuckwondo replied:
Actually, they're all there, but aliases are not exposed via the iterator. You'll see all of them in the [`__members__` attribute](https://docs.python.org/3/library/enum.html#enum.EnumType.__members__). However, if dealing with that is clumsy, I'm fine for you to stick with your other simplification.

ceholden replied:
Aah they indeed are there, thanks ✅. I wasn't getting them in autocomplete in my editor or in IPython so I assumed they weren't

chuckwondo replied:
Yeah, that's probably a deal-breaker then. Seems like keeping the uniqueness constraint affords a better devX.

ceholden commented:
Thanks for the close review @chuckwondo! I think I addressed pretty much everything directly except for the suggestions about making `lambda/` into a module. I'm hoping I addressed the issues those suggestions would address, but let me know if you're having issues with editor/IDE/etc

Rather than keeping `lambdas` as a package, I `git mv lambdas/ src/` and configured the `pyproject.toml` to look for packages under `src/`. I think this will resolve the issues with IDE tooling not finding the code while keeping the imports simpler. It at least worked for my editor, with features like "go to definition" or "go to references" working better than last week. I think this also dovetails better with the idea of having each subdirectory in `src/` be separate packages and treating this as a "workspace" of packages, which I'm keeping in mind for easier dependency management.

Let me know how this looks afterwards!

I also added a script to populate the `.env.[stage]` from our Github environment and some instructions for how to populate the `.env.[stage]` from either the template or this script.

chuckwondo commented on src/common/models.py:
```diff
@@ -0,0 +1,137 @@
+            version=".".join([version_major, version_minor]),
+        )
+
+    def to_str(self) -> str:
```
Consider renaming `to_str` to `__str__` and changing expressions elsewhere from `granule_id.to_str()` to `str(granule_id)`. This would also allow simpler usage in f-strings (`f"...{granule_id}..."` rather than `f"...{granule_id.to_str()}..."`).

ceholden replied:
Nice!

chuckwondo reviewed (approved):
Everything is looking fantastic! I pulled the latest changes and the devX works well while also being able to successfully use CDK.

Just a few minor suggestions, plus this one, which I couldn't comment on inline since it's not in the PR: In `scripts/test`, change `--cov=lambdas/` to `--cov=src/`.

Approving!

ceholden commented on cdk/stack.py:
```diff
@@ -13,16 +16,76 @@
+            groups_arg = " "
+
+        return [
+            "python -m venv uv_venv",
+            ". uv_venv/bin/activate",
+            "pip install uv",
+            "export UV_CACHE_DIR=/tmp",
+            f"cd {UV_ASSET_REQUIREMENTS}",
+            (
+                f"uv export {groups_arg} --frozen --no-emit-project --no-dev "
+                f"--no-default-groups --no-editable -o {input_dir}/requirements.txt"
+            ),
+            f"rm -rf {input_dir}/uv_venv",
```
I think +1 to using `input_dir` at the top, but I haven't found a good way to delete the `requirements.txt` this produces, even if I put the command in the "after bundling" 🤔

I'll give the "after bundling" another try.. from reading the code it seems like the "pip install" should happen _before_ the "after bundling" commands are run,
https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-lambda-python-alpha/lib/bundling.ts#L134

chuckwondo replied:
Oops, yeah looks like I should not have added removal of requirements.txt here. Oddly, I was still able to run `cdk diff`.

ceholden replied:
The CDK stuff will still succeed, but the Lambda won't have any of the required packages 😭. I think we could include a test phase as part of the `CommandHook`, but I'm more waiting on the official `uv` support to be merged and released to solve this for good 🙏

ceholden commented on tests/test_inventory_converter.py:
```diff
@@ -0,0 +1,78 @@
+
+import numpy as np
+import pandas as pd
+import pytest
```
Ah I was missing this from "known first party", fixed!
